File: src/game/Map.cpp
#include "nlohmann/json.hpp"
#include "sprite/Files.h"
#include "Map.h"

void trippin::Map::load(const std::string &name) {
    auto in = readFile(getMapFile(name).c_str());
    nlohmann::json j;
    in >> j;
    j.get_to(*this);
}

std::string trippin::Map::getMapFile(const std::string &name) {
    std::stringstream path;
    path << "maps/" << name << ".json";
    return path.str();
}

void trippin::Map::rescale(int scale) {
    universe *= scale;

    for (auto &obj: objects) {
        obj.position *= scale;
        obj.activation *= scale;
    }

    for (auto &layer: layers) {
        layer.size *= scale;
        for (auto &obj: layer.objects) {
            obj.position *= scale;
        }
    }
}

void trippin::from_json(const nlohmann::json &j, Map &map) {
    j.at("universe").at("width").get_to(map.universe.x);
    j.at("universe").at("height").get_to(map.universe.y);
    j.at("music").get_to(map.music);
    j.at("objects").get_to(map.objects);
    j.at("layers").get_to(map.layers);
}

void trippin::from_json(const nlohmann::json &j, Map::Object &obj) {
    j.at("id").get_to(obj.id);
    j.at("type").get_to(obj.type);
    j.at("position").get_to(obj.position);
    if (j.contains("lane"))
        j.at("lane").get_to(obj.lane);
    if (j.contains("sparkle"))
        j.at("sparkle").get_to(obj.sparkle);
    if (j.contains("activation"))
        j.at("activation").get_to(obj.activation);
}

void trippin::from_json(const nlohmann::json &j, Map::Layer &layer) {
    j.at("width").get_to(layer.size.x);
    j.at("height").get_to(layer.size.y);
    j.at("objects").get_to(layer.objects);
    if (j.contains("anchorTop"))
        j.at("anchorTop").get_to(layer.anchorTop);
}

void trippin::from_json(const nlohmann::json &j, Map::LayerObject &obj) {
    j.at("type").get_to(obj.type);
    j.at("position").get_to(obj.position);
}

File: tools/map.py
import xml.etree.ElementTree as ET
import json

namespace = {
    'svg': 'http://www.w3.org/2000/svg'
}


def extract_from_xml(node, name, target, fn):
    if name in node.attrib:
        target[name] = fn(node.attrib[name])


def find_objects(root):
    objects = []
    ints = ['lane', 'activation']
    bools = ['sparkle']
    id_counter = 1
    # an object is an image within a group that does not have attribute type='layer'
    for g in root.findall('.//svg:g', namespace):
        if g.attrib.get('type') not in ['layer', 'development']:
            for node in g.findall('.//svg:image', namespace):
                obj = {
                    'id': id_counter,
                    'type': node.attrib['type'],
                    'position': {
                        'x': int(node.attrib['x']),
                        'y': int(node.attrib['y'])
                    }
                }
                for s in ints:
                    extract_from_xml(node, s, obj, int)
                for s in bools:
                    extract_from_xml(node, s, obj, bool)
                if 'activateWhen' in node.attrib:
                    id = node.attrib['activateWhen']
                    ref = root.findall(f'.//*[@id="{id}"]')
                    if ref:
                        ref_x = int(ref[0].attrib['x'])
                        ref_activation = int(ref[0].attrib['activation'])
                        obj['activation'] = ref_activation + (obj['position']['x'] - ref_x)
                objects.append(obj)
                id_counter = id_counter + 1
    return objects


def find_layers(root):
    layers = []
    for g in root.findall('.//svg:g[@type="layer"]', namespace):
        layer = {
            'objects': []
        }
        if g.attrib.get('anchorTop') == 'true':
            layer['anchorTop'] = True
        layers.append(layer)
        for e in g.findall('.//svg:rect', namespace):
            layer['width'] = int(e.attrib['width'])
            layer['height'] = int(e.attrib['height'])
        for node in g.findall('.//svg:image', namespace):
            obj = {
                'type': node.attrib['type'],
                'position': {
                    'x': int(node.attrib['x']),
                    'y': int(node.attrib['y'])
                }
            }
            layer['objects'].append(obj)
    return layers


def make_model(svg_file):
    tree = ET.parse(svg_file)
    root = tree.getroot()
    return {
        'universe': {
            'width': int(root.attrib['width']),
            'height': int(root.attrib['height'])
        },
        'music': root.attrib['music'],
        'objects': find_objects(root),
        'layers': find_layers(root)
    }


def serialize_model(model, output_file):
    with open(output_file, 'w') as f:
        json.dump(model, f, indent=2)
